master/buildbot/test/fake/gce.py

```diff
@@ -0,0 +1,169 @@
+#
+# Copyright Buildbot Team Members
+
+from buildbot.util import gceclientservice
+
+
+class GCEAsyncResult:
+    def __init__(self, record_id):
+        self.record_id = record_id
+
+
+class ExpectedRequest:
+    NORMAL = 1
+    UNTIL = 2
+
+    def __init__(self, mode, httpMethod, url, params={}, json={}, result={}, data=None, resultCode=200):
+        self.mode = mode
+        self.httpMethod = httpMethod
+        self.url = url
+        self.params = params
+        self.json = json
+        self.data = data
+        self.result = result
+        self.resultCode = resultCode
+
+
+class GCERecorder(gceclientservice.GCEClientService):
+    IGNORE = object()
+
+    def __init__(self, scopes, sa_credentials, project=None, zone=None, instance=None, renderer=None):
+        self.expectations = []
+        self.record_id = 0
+        self.record = []
+        self.inFlight = {}
+        self.asyncId = 0
+
+        gceclientservice.GCEClientService.__init__(
+            self, scopes, sa_credentials, project=project, zone=zone, instance=instance,
+            renderer=renderer)
+
+        self._doRequest = self.interceptRequest
+
+    def expect(self, method, url, params={}, json={}, data=None, result={}, resultCode=200):
+        self.expectations.append(ExpectedRequest(
+            ExpectedRequest.NORMAL, method, url, params=params, json=json, data=data,
+            result=result, resultCode=resultCode))
+
+    def expectOperationRequest(self, method, url, params={}, json={}, data=None, resultCode=200):
+        self.asyncId += 1
+        selfLink = 'async-{}'.format(self.asyncId)
+        self.expectations.append(ExpectedRequest(
+            ExpectedRequest.NORMAL, method, url, params=params, json=json, data=data,
+            result={'selfLink': selfLink, 'status': 'STARTED'},
+            resultCode=resultCode))
+        return selfLink
+
+    def expectWaitForOperation(self, selfLink, status='DONE'):
+        self.expectations.append(ExpectedRequest(
+            ExpectedRequest.NORMAL, 'GET', selfLink,
+            result={'status': 'DONE'}, resultCode=200))
+
+    def expectInstanceStateWait(self, targetState):
+        return self.expect(
+            'GET', '/compute/v1/projects/p/zones/z/instances/i',
+            params={'fields': 'status'},
+            result={'status': targetState})
+
+    def ignoreUntil(self, method, url):
+        self.expectations.append(ExpectedRequest(
+            ExpectedRequest.UNTIL, method, url))
+
+    def compare(self, expected, actual):
+        if expected == GCERecorder.IGNORE:
+            return True
+        elif callable(expected):
+            return expected(actual)
+        else:
+            return expected == actual
+
+    def validateIsExpected(self, method, url, params, json, data):
+        if not self.expectations:
+            msg = "got {} {} but was not expecting any request".format(method, url)
+            assert False, msg
+
+        request = self.expectations[0]
```

Comment:
I think we could greatly simplify this method if we passed the test case to __init__, stored it as a member variable and then used `self.testcase.assertEqual()` and friends.